PetAdoptions/pethistory/index.js

```diff
@@ -0,0 +1,237 @@
+'use strict';
+
+const AWS = require('aws-sdk');
+const { Client } = require('pg');
+const Joi = require('joi');
+
+// Schema validation for SQS messages
+const adoptionMessageSchema = Joi.object({
+    transactionId: Joi.string().uuid().required(),
+    petId: Joi.string().required(),
+    petType: Joi.string().required(),
+    userId: Joi.string().required(),
+    adoptionDate: Joi.string().isoDate().required(),
+    timestamp: Joi.string().isoDate().required()
+});
+
+// Environment variables
+const {
+    RDS_SECRET_ARN,
+    AWS_REGION,
+} = process.env;
+
+// AWS clients
+const secretsManager = new AWS.SecretsManager({ region: AWS_REGION });
+
+/**
+ * Lambda handler for processing adoption history messages from SQS
+ */
+exports.handler = async (event, context) => {
+    console.log('Processing adoption history messages:', JSON.stringify(event, null, 2));
+
+    const results = [];
+    const batchItemFailures = [];
+
+    // Process each SQS record
+    for (const record of event.Records) {
+        try {
+            const result = await processAdoptionMessage(record);
+            results.push(result);
+        } catch (error) {
+            console.error('Failed to process record:', record.messageId, error);
+            // Add to batch failures instead of throwing
+            batchItemFailures.push({
+                itemIdentifier: record.messageId
+            });
+        }
+    }
+
+    console.log(`Processed ${results.length} messages successfully, ${batchItemFailures.length} failed`);
+
+    // Return batch failure information for SQS to handle partial failures
+    return {
+        batchItemFailures: batchItemFailures,
+        processedCount: results.length,
+        results: results
+    };
+};
+
+/**
+ * Process a single adoption history message from SQS
+ */
+async function processAdoptionMessage(record) {
+    const messageId = record.messageId;
+    // const receiptHandle = record.receiptHandle;
+
+    console.log(`Processing message ${messageId}`);
+
+    // Parse and validate message
+    let adoptionData;
+    try {
+        adoptionData = JSON.parse(record.body);
+        console.log('Parsed adoption data:', adoptionData);
+    } catch (error) {
+        console.error('Invalid JSON in message:', messageId, error);
+        throw new Error(`Invalid JSON in message ${messageId}: ${error.message}`);
+    }
+    // Validate message schema
+    const { error, value } = adoptionMessageSchema.validate(adoptionData);
+    if (error) {
+        console.error('Schema validation failed for message:', messageId, error.details);
+        throw new Error(`Schema validation failed for message ${messageId}: ${error.message}`);
+    }
+    const validatedData = value;
+    try {
+        console.log(`Processing adoption - Transaction: ${validatedData.transactionId}, Pet: ${validatedData.petId}, User: ${validatedData.userId}`);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when untrusted user input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input using `encodeURIComponent()` or `DOMPurify.sanitize()` before logging.  Learn more - https://cwe.mitre.org/data/definitions/117.html